File: orb/apps/ios/kagami-ios/KagamiIOS/Services/CommandRegistry.swift
//
// CommandRegistry.swift — Extensible Command System
//
// Provides a plugin-style architecture for:
// - Slash commands (/scene, /lights, etc.)
// - @ mentions (@room, @scene, etc.)
//
// Colony: Forge (e2) — Implementation
//

import Foundation

// MARK: - Command Protocol

@MainActor
protocol KagamiCommand {
    var name: String { get }
    var description: String { get }
    var aliases: [String] { get }
    var argsHint: String? { get }

    func execute(args: String) async throws -> CommandResult
    func suggest(partial: String) async -> [String]
}

extension KagamiCommand {
    var aliases: [String] { [] }
    var argsHint: String? { nil }
    func suggest(partial: String) async -> [String] { [] }
}

// MARK: - Command Result

struct CommandResult: @unchecked Sendable {
    let success: Bool
    let message: String?
    let data: [String: Any]?

    static func success(_ message: String? = nil) -> CommandResult {
        CommandResult(success: true, message: message, data: nil)
    }

    static func failure(_ message: String) -> CommandResult {
        CommandResult(success: false, message: message, data: nil)
    }
}

// MARK: - Mention Handler Protocol

protocol MentionHandler {
    var type: String { get }
    var prefix: String { get }

    func fetch(query: String) async -> [MentionItem]
    func resolve(item: MentionItem) -> [String: Any]
}

struct MentionItem: Identifiable, Hashable {
    let id: String
    let label: String
    let icon: String
    let value: String
    let secondary: String?

    init(id: String, label: String, icon: String = "", value: String? = nil, secondary: String? = nil) {
        self.id = id
        self.label = label
        self.icon = icon
        self.value = value ?? label
        self.secondary = secondary
    }
}

// MARK: - Command Registry

@MainActor
class CommandRegistry: ObservableObject {
    static let shared = CommandRegistry()

    @Published private(set) var commands: [String: any KagamiCommand] = [:]
    @Published private(set) var mentionHandlers: [String: any MentionHandler] = [:]

    private init() {
        registerBuiltinCommands()
        registerBuiltinMentions()
    }

    // MARK: - Registration

    func register(command: any KagamiCommand) {
        commands[command.name] = command
        for alias in command.aliases {
            commands[alias] = command
        }
    }

    func register(mentionHandler: any MentionHandler) {
        mentionHandlers[mentionHandler.type] = mentionHandler
    }

    // MARK: - Command Execution

    func parseSlashCommand(_ input: String) -> (command: (any KagamiCommand)?, args: String) {
        guard input.hasPrefix("/") else { return (nil, input) }

        let trimmed = String(input.dropFirst())
        let parts = trimmed.split(separator: " ", maxSplits: 1)
        let name = parts.first.map(String.init)?.lowercased() ?? ""
        let args = parts.count > 1 ? String(parts[1]) : ""

        return (commands[name], args)
    }

    func executeSlashCommand(_ input: String) async throws -> CommandResult {
        let (command, args) = parseSlashCommand(input)
        guard let command = command else {
            throw CommandError.unknownCommand(input.split(separator: " ").first.map(String.init) ?? input)
        }
        return try await command.execute(args: args)
    }

    // MARK: - Suggestions

    func getSuggestions(for input: String) async -> [CommandSuggestion] {
        guard input.hasPrefix("/") else { return [] }

        let trimmed = String(input.dropFirst())
        let parts = trimmed.split(separator: " ", maxSplits: 1)
        let partial = parts.first.map(String.init)?.lowercased() ?? ""

        // If we have a complete command and args, get arg suggestions
        if parts.count > 1, let command = commands[partial] {
            let argPartial = String(parts[1])
            let suggestions = await command.suggest(partial: argPartial)
            return suggestions.map {
                CommandSuggestion(
                    label: "/\(command.name) \($0)",
                    icon: "arrow.right",
                    value: "/\(command.name) \($0)"
                )
            }
        }

        // Get matching commands
        let uniqueCommands = Set(commands.values.map { $0.name })
        return uniqueCommands
            .filter { $0.hasPrefix(partial) }
            .sorted()
            .prefix(8)
            .map { name in
                let cmd = commands[name]!
                let hint = cmd.argsHint.map { " \($0)" } ?? ""
                return CommandSuggestion(
                    label: "/\(name)",
                    secondary: "\(hint) — \(cmd.description)",
                    icon: "chevron.right",
                    value: "/\(name) "
                )
            }
    }

    func getMentionSuggestions(type: String?, query: String) async -> [MentionItem] {
        if let type = type, let handler = mentionHandlers[type] {
            return await handler.fetch(query: query)
        }

        // Return mention type suggestions
        return mentionHandlers.keys.sorted().filter {
            query.isEmpty || $0.hasPrefix(query.lowercased())
        }.map { type in
            MentionItem(
                id: type,
                label: "@\(type)",
                icon: mentionHandlers[type]?.prefix ?? "@",
                value: "@\(type):",
                secondary: "Search \(type)s"
            )
        }
    }

    // MARK: - Built-in Commands

    private func registerBuiltinCommands() {
        register(command: SceneCommand())
        register(command: LightsCommand())
        register(command: TVCommand())
        register(command: FireplaceCommand())
        register(command: AnnounceCommand())
        register(command: HelpCommand(registry: self))
    }

    private func registerBuiltinMentions() {
        register(mentionHandler: RoomMentionHandler())
        register(mentionHandler: SceneMentionHandler())
    }
}

// MARK: - Suggestion Types

struct CommandSuggestion: Identifiable {
    let id = UUID()
    let label: String
    var secondary: String? = nil
    let icon: String
    let value: String
}

// MARK: - Errors

enum CommandError: LocalizedError {
    case unknownCommand(String)
    case invalidArgs(String)
    case executionFailed(String)

    var errorDescription: String? {
        switch self {
        case .unknownCommand(let cmd): return "Unknown command: \(cmd)"
        case .invalidArgs(let msg): return "Invalid arguments: \(msg)"
        case .executionFailed(let msg): return "Failed: \(msg)"
        }
    }
}

// MARK: - Built-in Commands

struct SceneCommand: KagamiCommand {
    let name = "scene"
    let description = "Execute a scene"
    let aliases = ["s"]
    let argsHint: String? = "<movie|goodnight|welcome>"

    func execute(args: String) async throws -> CommandResult {
        let scene = args.trimmingCharacters(in: .whitespaces).lowercased()
        guard !scene.isEmpty else {
            throw CommandError.invalidArgs("Scene name required")
        }
        await KagamiAPIService.shared.executeScene(scene)
        return .success("Scene \(scene) activated")
    }

    func suggest(partial: String) async -> [String] {
        let scenes = ["movie", "goodnight", "welcome", "exit_movie"]
        return scenes.filter { $0.hasPrefix(partial.lowercased()) }
    }
}

struct LightsCommand: KagamiCommand {
    let name = "lights"
    let description = "Set light level"
    let aliases = ["l", "light"]
    let argsHint: String? = "<0-100> [room]"

    func execute(args: String) async throws -> CommandResult {
        let parts = args.split(separator: " ")
        guard let levelStr = parts.first, let level = Int(levelStr) else {
            throw CommandError.invalidArgs("Level (0-100) required")
        }

        let room = parts.count > 1 ? parts.dropFirst().joined(separator: " ") : nil

        if let room = room {
            await KagamiAPIService.shared.setLights(level, rooms: [room])
        } else {
            await KagamiAPIService.shared.setLights(level)
        }

        return .success("Lights set to \(level)%")
    }

    func suggest(partial: String) async -> [String] {
        let levels = ["0", "25", "50", "75", "100"]
        if partial.isEmpty { return levels }

        // If partial starts with a number, suggest rooms
        if partial.first?.isNumber == true {
            let rooms = ["Living Room", "Kitchen", "Office", "Bedroom", "Dining"]
            let parts = partial.split(separator: " ", maxSplits: 1)
            if parts.count > 1 {
                let roomQuery = String(parts[1]).lowercased()
                return rooms
                    .filter { $0.lowercased().contains(roomQuery) }
                    .map { "\(parts[0]) \($0)" }
            }
        }

        return levels.filter { $0.hasPrefix(partial) }
    }
}

struct TVCommand: KagamiCommand {
    let name = "tv"
    let description = "Control TV mount"
    let argsHint: String? = "<up|down>"

    func execute(args: String) async throws -> CommandResult {
        let action = args.trimmingCharacters(in: .whitespaces).lowercased()
        guard ["up", "down", "raise", "lower"].contains(action) else {
            throw CommandError.invalidArgs("Use: /tv up or /tv down")
        }

        let normalizedAction = action == "up" ? "raise" : action == "down" ? "lower" : action
        await KagamiAPIService.shared.tvControl(normalizedAction)
        return .success("TV \(normalizedAction)")
    }

    func suggest(partial: String) async -> [String] {
        return ["up", "down"].filter { $0.hasPrefix(partial.lowercased()) }
    }
}

struct FireplaceCommand: KagamiCommand {
    let name = "fireplace"
    let description = "Toggle fireplace"
    let aliases = ["fire"]
    let argsHint: String? = "[on|off]"

    func execute(args: String) async throws -> CommandResult {
        let state = args.trimmingCharacters(in: .whitespaces).lowercased()
        let on = state != "off"
        await KagamiAPIService.shared.toggleFireplace(on: on)
        return .success("Fireplace \(on ? "on" : "off")")
    }

    func suggest(partial: String) async -> [String] {
        return ["on", "off"].filter { $0.hasPrefix(partial.lowercased()) }
    }
}

struct AnnounceCommand: KagamiCommand {
    let name = "announce"
    let description = "Announce message"
    let aliases = ["say", "tts"]
    let argsHint: String? = "<message>"

    func execute(args: String) async throws -> CommandResult {
        let message = args.trimmingCharacters(in: .whitespaces)
            .trimmingCharacters(in: CharacterSet(charactersIn: "\"'"))
        guard !message.isEmpty else {
            throw CommandError.invalidArgs("Message required")
        }
        await KagamiAPIService.shared.announce(message)
        return .success("Announced")
    }
}

struct HelpCommand: KagamiCommand {
    let name = "help"
    let description = "Show available commands"
    let aliases = ["?", "commands"]

    let registry: CommandRegistry

    func execute(args: String) async throws -> CommandResult {
        let commandList = Set(registry.commands.values.map { $0.name })
            .sorted()
            .map { name -> String in
                let cmd = registry.commands[name]!
                let hint = cmd.argsHint.map { " \($0)" } ?? ""
                return "/\(name)\(hint) — \(cmd.description)"
            }
            .joined(separator: "\n")

        return CommandResult(success: true, message: commandList, data: ["commands": commandList])
    }
}

// MARK: - Built-in Mention Handlers

struct RoomMentionHandler: MentionHandler {
    let type = "room"
    let prefix = "🏠"

    func fetch(query: String) async -> [MentionItem] {
        let rooms = [
            ("living", "Living Room", "1st"),
            ("kitchen", "Kitchen", "1st"),
            ("dining", "Dining", "1st"),
            ("office", "Office", "2nd"),
            ("bedroom", "Primary Bedroom", "2nd"),
            ("loft", "Loft", "2nd"),
            ("game", "Game Room", "Basement"),
            ("gym", "Gym", "Basement"),
        ]

        return rooms
            .filter { query.isEmpty || $0.1.lowercased().contains(query.lowercased()) }
            .map { MentionItem(id: $0.0, label: $0.1, icon: "🏠", secondary: $0.2) }
    }

    func resolve(item: MentionItem) -> [String: Any] {
        return ["room": item.id, "roomName": item.label]
    }
}

struct SceneMentionHandler: MentionHandler {
    let type = "scene"
    let prefix = "🎬"

    func fetch(query: String) async -> [MentionItem] {
        let scenes = [
            ("movie", "Movie Mode", "🎬"),
            ("goodnight", "Goodnight", "🌙"),
            ("welcome", "Welcome Home", "🏠"),
            ("exit_movie", "Exit Movie", "☀️"),
        ]

        return scenes
            .filter { query.isEmpty || $0.1.lowercased().contains(query.lowercased()) }
            .map { MentionItem(id: $0.0, label: $0.1, icon: $0.2) }
    }

    func resolve(item: MentionItem) -> [String: Any] {
        return ["scene": item.id]
    }
}

File: orb/apps/android/kagami-android/app/src/main/java/com/kagami/android/services/CommandRegistry.kt
/**
 * Kagami Command Registry — Extensible Command System
 *
 * Provides a plugin-style architecture for:
 * - Slash commands (/scene, /lights, etc.)
 * - @ mentions (@room, @scene, etc.)
 *
 * Colony: Forge (e2) — Implementation
 *
 * Architecture Note: Commands receive KagamiApiService via execute() parameter
 * to avoid direct KagamiApp.instance access. The registry is initialized
 * with the API service during app startup via DI.
 */

package com.kagami.android.services

import javax.inject.Inject
import javax.inject.Singleton

// MARK: - Command Interface

interface KagamiCommand {
    val name: String
    val description: String
    val aliases: List<String> get() = emptyList()
    val argsHint: String? get() = null

    suspend fun execute(args: String, apiService: KagamiApiService): CommandResult
    suspend fun suggest(partial: String): List<String> = emptyList()
}

// MARK: - Command Result

data class CommandResult(
    val success: Boolean,
    val message: String? = null,
    val data: Map<String, Any>? = null
) {
    companion object {
        fun success(message: String? = null) = CommandResult(true, message)
        fun failure(message: String) = CommandResult(false, message)
    }
}

// MARK: - Mention Handler Interface

interface MentionHandler {
    val type: String
    val prefix: String

    suspend fun fetch(query: String): List<MentionItem>
    fun resolve(item: MentionItem): Map<String, Any>
}

data class MentionItem(
    val id: String,
    val label: String,
    val icon: String = "",
    val value: String = label,
    val secondary: String? = null
)

// MARK: - Command Suggestion

data class CommandSuggestion(
    val label: String,
    val secondary: String? = null,
    val icon: String,
    val value: String
)

// MARK: - Command Registry

/**
 * Singleton command registry with proper DI support.
 * Commands receive the API service as a parameter to avoid global state access.
 */
@Singleton
class CommandRegistry @Inject constructor() {
    private val _commands = mutableMapOf<String, KagamiCommand>()
    val commands: Map<String, KagamiCommand> get() = _commands

    private val _mentionHandlers = mutableMapOf<String, MentionHandler>()
    val mentionHandlers: Map<String, MentionHandler> get() = _mentionHandlers

    init {
        registerBuiltinCommands()
        registerBuiltinMentions()
        // Register instance for static access
        _instance = this
    }

    companion object {
        @Volatile
        private var _instance: CommandRegistry? = null

        /**
         * Get singleton instance. Note: This should only be called after DI initialization.
         * For Composable functions, prefer using hiltViewModel or LocalComposition.
         */
        fun getInstance(): CommandRegistry = _instance ?: CommandRegistry().also { _instance = it }

        // Static convenience methods that delegate to instance
        suspend fun getSuggestions(input: String): List<CommandSuggestion> =
            getInstance().getSuggestions(input)

        suspend fun getMentionSuggestions(type: String?, query: String): List<MentionItem> =
            getInstance().getMentionSuggestions(type, query)

        suspend fun executeSlashCommand(input: String, apiService: KagamiApiService): CommandResult =
            getInstance().executeSlashCommand(input, apiService)
    }

    // MARK: - Registration

    fun register(command: KagamiCommand) {
        _commands[command.name] = command
        command.aliases.forEach { _commands[it] = command }
    }

    fun register(handler: MentionHandler) {
        _mentionHandlers[handler.type] = handler
    }

    // MARK: - Command Execution

    fun parseSlashCommand(input: String): Pair<KagamiCommand?, String> {
        if (!input.startsWith("/")) return null to input

        val trimmed = input.drop(1)
        val parts = trimmed.split(" ", limit = 2)
        val name = parts.firstOrNull()?.lowercase() ?: ""
        val args = parts.getOrNull(1) ?: ""

        return _commands[name] to args
    }

    suspend fun executeSlashCommand(input: String, apiService: KagamiApiService): CommandResult {
        val (command, args) = parseSlashCommand(input)
        return command?.execute(args, apiService)
            ?: CommandResult.failure("Unknown command: ${input.split(" ").first()}")
    }

    // MARK: - Suggestions

    suspend fun getSuggestions(input: String): List<CommandSuggestion> {
        if (!input.startsWith("/")) return emptyList()

        val trimmed = input.drop(1)
        val parts = trimmed.split(" ", limit = 2)
        val partial = parts.firstOrNull()?.lowercase() ?: ""

        // If we have args, get arg suggestions
        if (parts.size > 1) {
            val command = _commands[partial] ?: return emptyList()
            val argPartial = parts[1]
            return command.suggest(argPartial).map { suggestion ->
                CommandSuggestion(
                    label = "/${command.name} $suggestion",
                    icon = "→",
                    value = "/${command.name} $suggestion"
                )
            }
        }

        // Get matching commands
        return _commands.entries
            .distinctBy { it.value.name }
            .filter { it.key.startsWith(partial) }
            .sortedBy { it.key }
            .take(8)
            .map { (_, cmd) ->
                val hint = cmd.argsHint?.let { " $it" } ?: ""
                CommandSuggestion(
                    label = "/${cmd.name}",
                    secondary = "$hint — ${cmd.description}",
                    icon = "/",
                    value = "/${cmd.name} "
                )
            }
    }

    suspend fun getMentionSuggestions(type: String?, query: String): List<MentionItem> {
        if (type != null && _mentionHandlers.containsKey(type)) {
            return _mentionHandlers[type]!!.fetch(query)
        }

        // Return mention type suggestions
        return _mentionHandlers.keys.sorted()
            .filter { query.isEmpty() || it.startsWith(query.lowercase()) }
            .map { mentionType ->
                MentionItem(
                    id = mentionType,
                    label = "@$mentionType",
                    icon = _mentionHandlers[mentionType]?.prefix ?: "@",
                    value = "@$mentionType:",
                    secondary = "Search ${mentionType}s"
                )
            }
    }

    // MARK: - Built-in Commands

    private fun registerBuiltinCommands() {
        register(SceneCommand())
        register(LightsCommand())
        register(TVCommand())
        register(FireplaceCommand())
        register(AnnounceCommand())
        register(HelpCommand())
    }

    private fun registerBuiltinMentions() {
        register(RoomMentionHandler())
        register(SceneMentionHandler())
    }
}

// MARK: - Built-in Commands

class SceneCommand : KagamiCommand {
    override val name = "scene"
    override val description = "Execute a scene"
    override val aliases = listOf("s")
    override val argsHint = "<movie|goodnight|welcome>"

    override suspend fun execute(args: String, apiService: KagamiApiService): CommandResult {
        val scene = args.trim().lowercase()
        if (scene.isEmpty()) return CommandResult.failure("Scene name required")

        apiService.executeScene(scene)
        return CommandResult.success("Scene $scene activated")
    }

    override suspend fun suggest(partial: String): List<String> {
        val scenes = listOf("movie", "goodnight", "welcome", "exit_movie")
        return scenes.filter { it.startsWith(partial.lowercase()) }
    }
}

class LightsCommand : KagamiCommand {
    override val name = "lights"
    override val description = "Set light level"
    override val aliases = listOf("l", "light")
    override val argsHint = "<0-100> [room]"

    override suspend fun execute(args: String, apiService: KagamiApiService): CommandResult {
        val parts = args.trim().split(Regex("\\s+"))
        val level = parts.firstOrNull()?.toIntOrNull()
            ?: return CommandResult.failure("Level (0-100) required")

        val room = if (parts.size > 1) parts.drop(1).joinToString(" ") else null

        if (room != null) {
            apiService.setLights(level, listOf(room))
        } else {
            apiService.setLights(level)
        }

        return CommandResult.success("Lights set to $level%")
    }

    override suspend fun suggest(partial: String): List<String> {
        val levels = listOf("0", "25", "50", "75", "100")
        if (partial.isEmpty()) return levels

        // If partial starts with a number, suggest rooms
        if (partial.firstOrNull()?.isDigit() == true) {
            val rooms = listOf("Living Room", "Kitchen", "Office", "Bedroom", "Dining")
            val parts = partial.split(Regex("\\s+"), limit = 2)
            if (parts.size > 1) {
                val roomQuery = parts[1].lowercase()
                return rooms
                    .filter { it.lowercase().contains(roomQuery) }
                    .map { "${parts[0]} $it" }
            }
        }

        return levels.filter { it.startsWith(partial) }
    }
}

class TVCommand : KagamiCommand {
    override val name = "tv"
    override val description = "Control TV mount"
    override val argsHint = "<up|down>"

    override suspend fun execute(args: String, apiService: KagamiApiService): CommandResult {
        val action = args.trim().lowercase()
        if (action !in listOf("up", "down", "raise", "lower")) {
            return CommandResult.failure("Use: /tv up or /tv down")
        }

        val normalizedAction = when (action) {
            "up" -> "raise"
            "down" -> "lower"
            else -> action
        }

        apiService.tvControl(normalizedAction)
        return CommandResult.success("TV $normalizedAction")
    }

    override suspend fun suggest(partial: String): List<String> {
        return listOf("up", "down").filter { it.startsWith(partial.lowercase()) }
    }
}

class FireplaceCommand : KagamiCommand {
    override val name = "fireplace"
    override val description = "Toggle fireplace"
    override val aliases = listOf("fire")
    override val argsHint = "[on|off]"

    override suspend fun execute(args: String, apiService: KagamiApiService): CommandResult {
        val state = args.trim().lowercase()
        val on = state != "off"
        apiService.toggleFireplace(on)
        return CommandResult.success("Fireplace ${if (on) "on" else "off"}")
    }

    override suspend fun suggest(partial: String): List<String> {
        return listOf("on", "off").filter { it.startsWith(partial.lowercase()) }
    }
}

class AnnounceCommand : KagamiCommand {
    override val name = "announce"
    override val description = "Announce message"
    override val aliases = listOf("say", "tts")
    override val argsHint = "<message>"

    override suspend fun execute(args: String, apiService: KagamiApiService): CommandResult {
        val message = args.trim().trim('"', '\'')
        if (message.isEmpty()) return CommandResult.failure("Message required")

        apiService.announce(message)
        return CommandResult.success("Announced")
    }
}

class HelpCommand : KagamiCommand {
    override val name = "help"
    override val description = "Show available commands"
    override val aliases = listOf("?", "commands")

    override suspend fun execute(args: String, apiService: KagamiApiService): CommandResult {
        // Note: This accesses the static commands list, not requiring apiService
        val commandList = "Available commands:\n" +
            "/scene <name> - Execute a scene\n" +
            "/lights <0-100> [room] - Set light level\n" +
            "/tv <up|down> - Control TV mount\n" +
            "/fireplace [on|off] - Toggle fireplace\n" +
            "/announce <message> - Announce message\n" +
            "/help - Show this help"

        return CommandResult(true, commandList, mapOf("commands" to commandList))
    }
}

// MARK: - Built-in Mention Handlers

class RoomMentionHandler : MentionHandler {
    override val type = "room"
    override val prefix = "🏠"

    override suspend fun fetch(query: String): List<MentionItem> {
        val rooms = listOf(
            Triple("living", "Living Room", "1st"),
            Triple("kitchen", "Kitchen", "1st"),
            Triple("dining", "Dining", "1st"),
            Triple("office", "Office", "2nd"),
            Triple("bedroom", "Primary Bedroom", "2nd"),
            Triple("loft", "Loft", "2nd"),
            Triple("game", "Game Room", "Basement"),
            Triple("gym", "Gym", "Basement"),
        )

        return rooms
            .filter { query.isEmpty() || it.second.lowercase().contains(query.lowercase()) }
            .map { MentionItem(it.first, it.second, "🏠", secondary = it.third) }
    }

    override fun resolve(item: MentionItem) = mapOf("room" to item.id, "roomName" to item.label)
}

class SceneMentionHandler : MentionHandler {
    override val type = "scene"
    override val prefix = "🎬"

    override suspend fun fetch(query: String): List<MentionItem> {
        val scenes = listOf(
            Triple("movie", "Movie Mode", "🎬"),
            Triple("goodnight", "Goodnight", "🌙"),
            Triple("welcome", "Welcome Home", "🏠"),
            Triple("exit_movie", "Exit Movie", "☀️"),
        )

        return scenes
            .filter { query.isEmpty() || it.second.lowercase().contains(query.lowercase()) }
            .map { MentionItem(it.first, it.second, it.third) }
    }

    override fun resolve(item: MentionItem) = mapOf("scene" to item.id)
}
